Tests: allow only validating changed YAML files

harding opened the pull request:
This makes `make` default to only JSON validate changed YAML files for the compatibility and topic files.  For everyday work on those files, this is a huge speed up in the write-test-debug loop.  Previously, a singe file change could take up to 10 seconds to generate an error; now such an error is returned is about 0.1 seconds.

For a fresh checkout by CI or if you run `make clean ; make`, all YAML files will still be validated.

This change also incidentally allows running the validator on multiple files concurrently with, e.g., via `make -j4`, although I wouldn't suggest that for production use.  I quickly fixed some concurrency issues in our makefile, but I'm sure there's race conditions remaining.  Still, it's probably acceptable for slightly faster testing during development if you run a no-concurrency test at the end (or rely on CI).

A separate commit also addresses this previous review comment: https://github.com/bitcoinops/bitcoinops.github.io/pull/555#discussion_r645485171

jnewbery commented:
Concept ACK

This seems to work mostly, but the bitcoin-core-wallet.yaml file is always revalidated for me.

First run:

```
→ make
bundle exec _contrib/schema-validator.rb _data/schemas/compatibility.yaml _data/compatibility/blockchaininfo.yaml
bundle exec _contrib/schema-validator.rb _data/schemas/compatibility.yaml _data/compatibility/bitgo.yaml
[...]
bundle exec _contrib/schema-validator.rb _data/schemas/compatibility.yaml _data/compatibility/brd.yaml
bundle exec _contrib/schema-validator.rb _data/schemas/topics.yaml _topics/en/output-script-descriptors.md
bundle exec _contrib/schema-validator.rb _data/schemas/topics.yaml _topics/en/replace-by-fee.md
[...]
bundle exec _contrib/schema-validator.rb _data/schemas/topics.yaml _topics/en/hold-invoices.md
bundle exec _contrib/schema-validator.rb _data/schemas/topics.yaml _topics/en/splicing.md
## Check for Markdown formatting problems
bundle exec mdl -g -r MD009 .
## Check that posts declare a slug, see issue #155 and PR #156
! git --no-pager grep -L "^slug: " _posts
[...]
```

Second run:

```
→ make
bundle exec _contrib/schema-validator.rb _data/schemas/compatibility.yaml _data/compatibility/bitcoin-core-wallet.yaml
## Check for Markdown formatting problems
bundle exec mdl -g -r MD009 .
## Check that posts declare a slug, see issue #155 and PR #156
! git --no-pager grep -L "^slug: " _posts
[...]
```

Third run:

```
→ make
bundle exec _contrib/schema-validator.rb _data/schemas/compatibility.yaml _data/compatibility/bitcoin-core-wallet.yaml
## Check for Markdown formatting problems
[...]
```

harding commented:
@jnewbery whoops, I saw that but forgot to fix it.  Fixed now, it was a file that didn't match our pattern; I renamed it (no change to the names of output files / all links still work).

bitschmidty commented:
tACK - was able to execute clean, production build, preview, and newsletter/email and all behaved as expected
